pkg/query-service/app/opamp/logspipeline_test.go

```diff
@@ -0,0 +1,161 @@
+package opamp
+
+import (
+	"fmt"
+	"testing"
+
+	. "github.com/smartystreets/goconvey/convey"
+	"go.signoz.io/signoz/pkg/query-service/constants"
+)
+
+var BuildProcessorTestData = []struct {
+	Name              string
+	agentConf         map[string]interface{}
+	pipelineProcessor map[string]interface{}
+	outputConf        map[string]interface{}
+}{
+	{
+		Name: "Add",
+		agentConf: map[string]interface{}{
+			"processors": map[string]interface{}{
+				"batch": struct{}{},
+			},
+		},
+		pipelineProcessor: map[string]interface{}{
+			constants.LogsPPLPfx + "_b": struct{}{},
+		},
+		outputConf: map[string]interface{}{
+			"processors": map[string]interface{}{
+				constants.LogsPPLPfx + "_b": struct{}{},
+				"batch":                     struct{}{},
+			},
+		},
+	},
+	{
+		Name: "Remove",
+		agentConf: map[string]interface{}{
+			"processors": map[string]interface{}{
+				constants.LogsPPLPfx + "_b": struct{}{},
+				"batch":                     struct{}{},
+			},
+		},
+		pipelineProcessor: map[string]interface{}{},
+		outputConf: map[string]interface{}{
+			"processors": map[string]interface{}{
+				"batch": struct{}{},
+			},
+		},
+	},
+	{
+		Name: "remove and upsert 1",
+		agentConf: map[string]interface{}{
+			"processors": map[string]interface{}{
+				constants.LogsPPLPfx + "_a": struct{}{},
+				constants.LogsPPLPfx + "_b": struct{}{},
+				"batch":                     struct{}{},
+			},
+		},
+		pipelineProcessor: map[string]interface{}{
+			constants.LogsPPLPfx + "_b": struct{}{},
+		},
+		outputConf: map[string]interface{}{
+			"processors": map[string]interface{}{
+				constants.LogsPPLPfx + "_b": struct{}{},
+				"batch":                     struct{}{},
+			},
+		},
+	},
+	{
+		Name: "remove and upsert 2",
+		agentConf: map[string]interface{}{
+			"processors": map[string]interface{}{
+				"memory_limiter":            struct{}{},
+				constants.LogsPPLPfx + "_a": struct{}{},
+				constants.LogsPPLPfx + "_b": struct{}{},
+				"batch":                     struct{}{},
+			},
+		},
+		pipelineProcessor: map[string]interface{}{
+			constants.LogsPPLPfx + "_b": struct{}{},
+		},
+		outputConf: map[string]interface{}{
+			"processors": map[string]interface{}{
+				"memory_limiter":            struct{}{},
+				constants.LogsPPLPfx + "_b": struct{}{},
+				"batch":                     struct{}{},
+			},
+		},
+	},
+}
+
+func TestBuildLogParsingProcessors(t *testing.T) {
+	for _, test := range BuildProcessorTestData {
+		Convey(test.Name, t, func() {
+			err := BuildLogParsingProcessors(test.agentConf, test.pipelineProcessor)
+			So(err, ShouldBeNil)
+			So(test.agentConf, ShouldResemble, test.outputConf)
+		})
+	}
+
+}
+
+var BuildLogsPipelineTestData = []struct {
+	Name             string
```

Comment:
I would love to have some eyes on these test cases.

Reply:
Yes so  the array of interfaces is nothing but an array of strings.

So my expected behaviour is that 
* if there is no existing `log parsing processor` in the processor array i.e `currentPipeline` add it to the log parsing processors to the beginning.
 eg:- currentPipeline is ` filter, batch`  and I want to add `logparser_a , logparser_b` then the final result is  `logparser_a , logparser_b, filter, batch`
 
 * If someone modifies the currentPipeline to `filter, logparser_a , logparser_b, batch` and the new version of log parsing pipeline contains only `logparser_a` then final result is `filter, logparser_a , batch`

* If someone adds a processor in between and currentPipeline is `filter, logparser_a , processor2, logparser_b, batch` and log parsing pipelines contains ` logparser_a , logparser_b, logparser_c` then `logparser_c` will be added by taking `logparser_b` as anchor the final result is `filter, logparser_a , processor2, logparser_b, logparser_c, batch`

Reply:
Okay, I am checking your examples, and I have updated the interfaces to structs.

Reply:
```
{
Name:             "rearrange pipelines",
currentPipeline:  []string{"processor1", "processor2", constants.LogsPPLPfx + "_a", "processor3", constants.LogsPPLPfx + "_b", "batch"},
logsPipeline:     []string{constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a"},
expectedPipeline: []string{"processor1", "processor2", "processor3", constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", "batch"},
},
```

Here a should are present next to b taking b as an anchor

Reply:
```
{
Name:             "rearrange pipelines with new processor",
currentPipeline:  []string{"processor1", "processor2", constants.LogsPPLPfx + "_a", "processor3", constants.LogsPPLPfx + "_b", "batch"},
logsPipeline:     []string{constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_c"},
expectedPipeline: []string{"processor1", "processor2", "processor3", constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_c", "batch"},
},
```

Here also we should take `_b` as the anchor and add the remaining ones.

Reply:
```
{
Name:             "last to first",
currentPipeline:  []string{"processor1", "processor2", constants.LogsPPLPfx + "_a", "processor3", "processor4", constants.LogsPPLPfx + "_b", "batch", constants.LogsPPLPfx + "_c"},
logsPipeline:     []string{constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_b"},
expectedPipeline: []string{"processor1", "processor2", "processor3", "processor4", "batch", constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_b"},
},
```
Take `_c` as anchor

Reply:
```
{
Name:             "multiple rearrange pipelines",
currentPipeline:  []string{"processor1", "processor2", constants.LogsPPLPfx + "_a", "processor3", constants.LogsPPLPfx + "_b", "batch", constants.LogsPPLPfx + "_c", "processor4", "processor5", constants.LogsPPLPfx + "_d", "processor6", "processor7"},
logsPipeline:     []string{constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_d", constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_e"},
expectedPipeline: []string{"processor1", "processor2", "processor3", constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", "batch", "processor4", "processor5", constants.LogsPPLPfx + "_d", constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_e", "processor6", "processor7"},
},
```

here also anchor should be taken on what is found if something is left then add it after it.

Reply:
```
{
Name:             "multiple rearrange with new pipelines",
currentPipeline:  []string{"processor1", "processor2", constants.LogsPPLPfx + "_a", "processor3", constants.LogsPPLPfx + "_b", "batch", constants.LogsPPLPfx + "_c", "processor4", "processor5", constants.LogsPPLPfx + "_d", "processor6", "processor7"},
logsPipeline:     []string{constants.LogsPPLPfx + "_z", constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_d", constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_e"},
expectedPipeline: []string{constants.LogsPPLPfx + "_z", "processor1", "processor2", "processor3", constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", "batch", "processor4", "processor5", constants.LogsPPLPfx + "_d", constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_e", "processor6", "processor7"},
},
```

Here also we go in order of `logsPipeline` and then if something is not in order we unmark it and add it later